apollo-router/src/configuration/mod.rs

```diff
@@ -603,6 +617,55 @@ pub(crate) struct RedisCache {
     pub(crate) urls: Vec<String>,
 }
 
+/// Configuration options pertaining to the subgraph server component.
+#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
+#[serde(deny_unknown_fields)]
+pub(crate) struct Tls {
+    #[serde(default)]
+    pub(crate) all: Subgraph,
+    #[serde(default)]
+    pub(crate) subgraphs: HashMap<String, Subgraph>,
+}
```

Comment:
What if later we decide to enable tls directly on supergraph ? This configuration might be wrong. Maybe it's too early to consider this but we could follow the same configuration shape than in other places with something like:

```yaml
tls:
    subgraph:
        all:
            ...
        subgraphs:
             user:
                 ...
    supergraph:
        ...
```

Maybe it's too early or useless to think about this now but ....

Reply:
I agree, we might add TLS on the frontend later. The issue here is that the traffic shaping plugin arranged to fields in another way:
https://www.apollographql.com/docs/router/configuration/traffic-shaping#configuration

```yaml
traffic_shaping:
  deduplicate_variables: true # Enable the variable deduplication optimization.
  router: # Rules applied to requests from clients to the router
    global_rate_limit: # Accept a maximum of 10 requests per 5 secs. Excess requests must be rejected.
      capacity: 10
      interval: 5s # Must not be greater than 18_446_744_073_709_551_615 milliseconds and not less than 0 milliseconds
    timeout: 50s # If a request to the router takes more than 50secs then cancel the request (30 sec by default)
  all:
    deduplicate_query: true # Enable query deduplication for all subgraphs.
    compression: br # Enable brotli compression for all subgraphs.
  subgraphs: # Rules applied to requests from the router to individual subgraphs
    products:
      deduplicate_query: false
```

I think it would make more sense to have `router` and `subgraph` sections at the top level, like you proposed, but then that would be either inconsistent with traffic shaping, or a config breaking change.
@BrynCooke does this look like something that could be fixed automatically with the config migrations?

Reply:
I think it should not be part of this PR but we definitely need to polish the traffic_shaping configuration because it's inconsistent with what we have currently. Thoughts @BrynCooke ?

Reply:
Some of this could be fixed with migrations, but agree that this needs more thought. I worry long term about ambiguous migrations due to us flip flopping. This is exasperated by the router config not having a version number.

Maybe we should create a skeleton plugin shape that we agree on up front? Let's bring proposals to the team.

```yaml
plugin:
  <global_config>
  router:
    <router_config>
    request:
       <router_request_config>  
    response:
       <router_response_config> 
  subgraphs:
    all:
      <all_subgraph_config>  
      request:
        <all_subgraph_request_config>  
      response:
        <all_subgraph_response_config>  
    products:
      <product_subgraph_config>  
      request:
        <product_subgraph_request_config>  
      response:
        <product_subgraph_response_config>  
```

Reply:
If router config did have a version number we would definitely be able to fix all the things via migrations.

Reply:
Agree cleanup should not be part of this pr

Reply:
alright, then I'll go with:

```yaml
tls:
    subgraph:
        all:
            ...
        subgraphs:
             user:
                 ...
```

and we'll follow up for various configuration changes later

Reply:
done in [93fa07e](https://github.com/apollographql/router/pull/2008/commits/93fa07eb29f195605083ab505f66e8942fdaec88)